ee/desktop/runtime/runner.go

```diff
@@ -109,3 +110,68 @@ func (r *DesktopUsersProcessesRunner) Interrupt(interruptError error) {
 		}
 	}
 }
+
+// waitForProcess adds 1 to DesktopUserProcessRunner.procsWg and runs a goroutine to wait on the process to exit.
+// The go routine will decrement DesktopUserProcessRunner.procsWg when it exits
+func (r *DesktopUsersProcessesRunner) waitOnProcessAsync(uid string, proc *os.Process) {
+	r.procsWg.Add(1)
+	go func(username string, proc *os.Process) {
+		defer r.procsWg.Done()
+		// if the desktop process dies, the parent must clean up otherwise we get a zombie process
+		// waiting here gives the parent a chance to clean up
+		_, err := proc.Wait()
+		if err != nil {
+			level.Error(r.logger).Log(
+				"msg", "desktop process died",
+				"uid", uid,
+				"pid", proc.Pid,
+				"err", err,
+			)
+		}
+	}(uid, proc)
+}
+
+// determineExecutablePath returns DesktopUsersProcessesRunner.executablePath if it is set,
+// otherwise it returns the path to the current binary.
+func (r *DesktopUsersProcessesRunner) determineExecutablePath() (string, error) {
+	if r.executablePath != "" {
+		return r.executablePath, nil
+	}
+
+	executable, err := os.Executable()
+	if err != nil {
+		return "", fmt.Errorf("error getting executable path: %w", err)
+	}
+
+	return executable, nil
+}
+
+func (r *DesktopUsersProcessesRunner) userHasDesktopProcess(uid string) bool {
+	// have no record of process
+	proc, ok := r.uidProcs[uid]
+	if !ok {
+		return false
+	}
+
+	// have a record of process, but it died for some reason, log it
+	if !processExists(proc.Pid) {
+		level.Info(r.logger).Log(
+			"msg", "found existing desktop process dead for console user",
+			"dead_pid", r.uidProcs[uid].Pid,
+			"uid", uid,
+		)
+
+		return false
+	}
+
+	// have running process
+	return true
+}
+
+func processExists(pid int) bool {
```

Comment:
What's the chance of pid reuse?

Reply:
I don't know. I updated the func to take `*os.Process` as a param in hopes that `os.Process` tracks pid, otherwise I'm not sure how one would prevent or test for process reuse .... are you thinking we should go down the route of looking at some of the process attributes (like path?) to make sure it's actually a launcher desktop process?

Reply:
Old skool would be to check the path (or name), and assume it's good enough to head off a reuse issue. Probably won't be an issue for end user devices, but it was common for servers.

Reply:
well .... I cannot imagine how difficult it would be to debug a pid reuse issue without having a terminal to the machine so I went ahead and added some logic to track and compare the path as well